Implemented to parse properly regarding resource bundle usage of C files

gouniLee opened the pull request:
Many of webOS services are written c or cpp program. I create this module to extract the localization string properly in C files. Here's API spec & example
```
const char* resBundle_getLocString(ResBundleC * bundle, const char * source)
const char* resBundle_getLocStringWithKey(ResBundleC * bundle, const char * key, const char * source)
```
```
char *screen_share_58 = (char *)resBundle_getLocString(res_bundle, "OK");
localized_arg_string = (gchar*)resBundle_getLocStringWithKey(_g_res_bundle_object, "pictureMode", "Standard");
```

Supporting i18n comment style(Not allowed multi-line comment) :
```
/* i18n  Button style */ 
// i18n  /** Connect WiSA Dongle **/ 
// i18n : Power button
// i18n : GUIDE button for Set-top box, used in screen remote
```
In this PR, I've done to implement features, wrote  test cases and make them passed all.

ehoogerbeets commented on CFile.js:
```diff
@@ -0,0 +1,229 @@
+    var unescaped = string;
+
+    unescaped = unescaped.
+        replace(/\\\\n/g, "").                // line continuation
```
I don't think C support line continuation, does it?

gouniLee replied:
I removed it. When making a key, We don't manipulate string at all. Regarding that, I added more test cases. String with a different number of spaces will be handled as a different key each. https://github.com/iLib-js/ilib-loctool-webos-c/blob/initialSetup/test/testCFile.js#L83

ehoogerbeets commented on CFile.js:
```diff
@@ -0,0 +1,229 @@
+    return CFile.unescapeString(source);
+};
+
+var reGetLocString = new RegExp(/\bresBundle_getLocString\((.*)\,\s*[\"|\'](.*)*[\"|\']\)\;/g);
```
This will not handle the case where there is an escaped double-quote character in the middle of the string. Instead, use a regular expression like:

```
"((\\"|[^"])*)"
```

to catch those.

ehoogerbeets replied:
Also, C does not support single-quoted strings.

gouniLee replied:
I remove all the test cases with single-quote strings.
and I don't get that case. (escaped double-quote character in the middle of the string) Could you give me an example? Do you mean the following case?  https://github.com/iLib-js/ilib-loctool-webos-c/blob/initialSetup/test/testCFile.js#L237

ehoogerbeets replied:
Yes, like that.

gouniLee replied:
look like current Regular expression is covering it. don't you think so?

ehoogerbeets replied:
The case where there are two strings to extract on the same line will make this regular expression fail.

```
resBundle_getLocString(foo, "bar"); resBundle_getLocString(asdf, "xxx");
```

Will return a source string `bar"); resBundle_getLocString(asdf, "xxx`

gouniLee replied:
I added a testcase to verify this feature.https://github.com/iLib-js/ilib-loctool-webos-c/blob/initialSetup/test/testCFile.js#L620

ehoogerbeets commented on CFile.js:
```diff
@@ -0,0 +1,229 @@
+
+var reGetLocString = new RegExp(/\bresBundle_getLocString\((.*)\,\s*[\"|\'](.*)*[\"|\']\)\;/g);
+var reGetLocStringWithKey = new RegExp(/\bresBundle_getLocStringWithKey\((.*)\,\s*[\"|\'](.*)[\"|\']\,\s*[\"|\'](.*)*[\"|\']\)/g);
+var reI18nComment = new RegExp("//\\s*i18n\\s*\\s*(.*)$");
```
C does not support double-slash comments. It only supports /* this style */ of comments.

gouniLee replied:
Yeah.. but we have cases to use double-slash in i18n comments. I think I need to support follow three styles.
```
1. // i18n comments hello
2. /* i18n comments hello */
3. /* i18n comments
    *  hello
    */
```

ehoogerbeets replied:
The double-slash comments should not appear in valid C files. It's probably okay to leave the support for it, but it is not needed.

gouniLee replied:
I've been thinking to support multi-line i18n comments, It's a bit weird. I think only a single-line comment would be fine. What do you think?
and I checked some C service codes in webOS. most of the comments were written double-slash style. That comment form doesn't occur any logic problem, and there wasn't a specific guideline so far, so I assume they just following JS i18n comment style..
They were some examples that found:
```
// i18n  /** Connect WiSA Dongle **/ 
// i18n : Power button
// i18n : GUIDE button for Set-top box, used in screen remote
```

ehoogerbeets commented on package.json:
```diff
@@ -0,0 +1,60 @@
+        "translation",
+        "webpack",
+        "loader",
+        "javascript"
```
change the keyword to "c" instead of "javascript". Also, we don't need "loader" or "webpack"

gouniLee replied:
I updated it.

ehoogerbeets commented on package.json:
```diff
@@ -0,0 +1,60 @@
+        "log4js": "^5.0.0"
+    },
+    "devDependencies": {
+        "loctool": "^2.2.0",
```
Should be "^2.4.0"

gouniLee replied:
I updated it.

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+        });
+        test.ok(cf);
+        
+        cf.parse("resBundle_getLocString(res_bundle, 'OK');");
```
C doesn't support single quote strings

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+        });
+        test.ok(cf);
+        
+        cf.parse("char *screen_share_58 = (char *)resBundle_getLocString(res_bundle, 'This is a test');");
```
Same here

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+        test.done();
+    },
+
+    testCFileParseMoreComplexSingleQuotes: function(test) {
```
Probably should not be named "SingleQuotes" when it doesn't test the single quotes

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+        });
+        test.ok(cf);
+
+        cf.parse('char* alert_btn= (char *)resBundle_getLocString(notification_getResBundle(), "OK"); // i18n button');
```
Change to /* this style */ of comments

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+            reskey: "PictureMode.Standard"
+        });
+        test.ok(r);
+        test.equal(r[0].getSource(), "PictureMode.Standard");
```
Source should be "Standard" not "PictureMode.Standard"

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+        });
+        test.ok(cf);
+
+        cf.parse('const char* pPicture = resBundle_getLocStringWithKey(resBundle,"PictureMode.Standard", "Standard"); // i18n button');
```
Fix /* comment */

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+            reskey: "PictureMode.Standard"
+        });
+        test.ok(r);
+        test.equal(r[0].getSource(), "PictureMode.Standard");
```
"Standard"

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+    },
+
+    testCFileParseWithKeySingleQuotes: function(test) {
+        test.expect(5);
```
Remove

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+    },
+
+    testCFileParseJSWithKeySingleQuotes: function(test) {
+        test.expect(5);
```
Remove

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+            reskey: "PictureMode.Standard"
+        });
+        test.ok(r);
+        test.equal(r[0].getSource(), "PictureMode.Standard");
```
"Standard"

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+            reskey: "Block.key"
+        });
+        test.ok(r);
+        test.equal(r[0].getSource(), "Block.key");
```
"Block"

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+            reskey: "Cancel.key"
+        });
+        test.ok(r);
+        test.equal(r[0].getSource(), "Cancel.key");
```
"Cancel"

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+        });
+        test.ok(cf);
+
+        cf.parse('char *screen_share_60 = (char *)resBundle_getLocStringWithKey(res_bundle, "Block.key", "Block");\n\ta.parse("This is another test.");\n\t\tchar *screen_share_67 = (char *)resBundle_getLocString(res_bundle, "Cancel"); \n\tchar *screen_share_70 = (char *)resBundle_getLocString(res_bundle, "Stop");');
```
These are on different lines (with a \n in the middle). The point of testing on the same line is to make sure that the regular expression can work properly with multiple strings to extract on the same line.

ehoogerbeets replied:
This one will still fail if you remove the \n in the parsed string. The regular expressions are not correct yet.

gouniLee replied:
I have additional test cases and it seems fine. Did I get it wrong?; ;(
https://github.com/iLib-js/ilib-loctool-webos-c/blob/initialSetup/test/testCFile.js#L620
https://github.com/iLib-js/ilib-loctool-webos-c/blob/initialSetup/test/testCFile.js#L654https://github.com/iLib-js/ilib-loctool-webos-c/blob/initialSetup/test/testCFile.js#L689

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+        });
+        test.ok(cf);
+
+        cf.parse('char *screen_share_60 = (char *)resBundle_getLocString(res_bundle, "Block");// i18n foo\n\ta.parse("This is another test.");\n\t\tchar *screen_share_67 = (char *)resBundle_getLocString(res_bundle, "Cancel");  // i18n bar');
```
Fix comments

gouniLee replied:
I need to support that comment style.

ehoogerbeets commented on test/testCFile.js:
```diff
@@ -0,0 +1,660 @@
+            "    struct json_object *jobj=NULL, *subjobj=NULL, *btnjobj=NULL, *pramjobj=NULL, *aryjobj=NULL, *closejobj=NULL;\n" +
+            "    char msg[1024]={0,};\n" +
+            "\n" +
+            "    char* alert_btn= (char *)resBundle_getLocString(notification_getResBundle(), 'OK');\n" +
```
Put C strings in double quotes

gouniLee replied:
I updated it.

ehoogerbeets commented on test/testCFileType.js:
```diff
@@ -0,0 +1,84 @@
+
+        var cft = new CFileType(p);
+        test.ok(cft);
+        test.ok(!cft.handles("foojs"));
```
Make that "fooc" instead. ie. that is "foo.c" without the dot

gouniLee replied:
I updated it.

ehoogerbeets commented on test/testfiles/t1.c:
```diff
@@ -0,0 +1,126 @@
+	char *screen_share_37 = (char *)resBundle_getLocString(res_bundle, "Please try again after closing the Screen Share app.");
+	char *screen_share_40 = (char *)resBundle_getLocString(res_bundle, "Overlay Mode is not available during recording or Live Playback.");
+	char *screen_share_41 = (char *)resBundle_getLocString(res_bundle, "Overlay Mode will be off now to start recording or Live Playback.");
+	char *screen_share_42 = (char *)resBundle_getLocString(res_bundle, "Overlay Mode is not available in {name}.");
```
Put escaped double quotes here somewhere to test that it parses properly

gouniLee commented:
I've updated code and added/fixed them corresponding to all your feedback. All test is passed including a double quote in String case. and Regarding i18n comment style I need to consider various cases with // i18n Please re-review them again. Thank you!

gouniLee commented on CFile.js:
```diff
@@ -42,69 +42,19 @@ var CFile = function(props) {
  * @returns {String} a unique key for this string
  */
 CFile.prototype.makeKey = function(source) {
-    return CFile.unescapeString(source);
```
I added the escaped function https://github.com/iLib-js/ilib-loctool-webos-c/blob/initialSetup/CFile.js#L52 so something like `\n or \t etc` is working. and I didn't wrap it when making a key. so I updated that part now. Regarding multi spaces in string, I'm not going to compress white space. so That's why only I added an `unescapedString` function.

gouniLee replied:
oh sorry.... ;(   I thought with `testCFileParseCSimple4` case,  it's ok. but I 've realized it doesn't work when I wrote  testcase in `t2.c` file.. It needs to handle something like `\n`\t` in strings. I'll re-update it..
